Prune the Term and Index in-memory storage.

jimwebber opened the pull request:
Previously kept all terms and indexes forever. Now we prune them when
the log is pruned.

jimwebber committed:
Prune the Term and Index in-memory storage.

Previously kept all terms and indexes forever. Now we prune them when
the log is pruned.

martinfurmanski commented on enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/log/segmented/Terms.java:
```diff
@@ -107,7 +107,30 @@ synchronized void truncate( long fromIndex )
+    {
+        for ( int i = 0; i < indexes.length; i++ )
+        {
+            if ( indexes[i] == min )
```
The indexes define ranges, so you can't have an exact comparison. You need to find the lowest i with index[i] >= min and then prune out those that precede it, 0..(i - 1).

martinfurmanski commented on enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/log/segmented/Terms.java:
```diff
@@ -107,7 +107,30 @@ synchronized void truncate( long fromIndex )
+            return;
+        }
+
+        size = 1;
```
This is the size of the array, you cannot reset it to 1. You should set it to the size of the remainder of the array.

jimwebber replied:
Done. Though having an array of size 1 did pass the tests.

martinfurmanski commented on enterprise/core-edge/src/test/java/org/neo4j/coreedge/raft/log/segmented/TermsTest.java:
```diff
@@ -276,16 +276,35 @@ public void shouldAppendAfterSkip() throws Exception
     }
 
+    @Test
+    public void shouldMaintainSaneMemoryUseForTerms() throws Exception
```
You could expose the size for the benefit of the test, and assert on it shrinking. That should lead to tests that are more on-point.

jimwebber replied:
I've gone with reflection here, but now assert on size rather than on not blowing up.

martinfurmanski replied:
You still need to assert on the actual terms, i.e. which terms are expected to still exist and which are expected to have been pruned out. Would be good to test a few different scenarios, for example pruning nothing, pruning everything, pruning on the boundary of a range, pruning in the middle of a range, pruning a few ranges.

martinfurmanski commented on enterprise/core-edge/src/test/java/org/neo4j/coreedge/raft/log/segmented/TermsTest.java:
```diff
@@ -276,16 +276,35 @@ public void shouldAppendAfterSkip() throws Exception
+        terms = new Terms( 1, 1 );
+
+        // when
+        for ( int i = 2; i < 1_000_000; i++ )
```
1M \* 64 \* 2 is still just 128MB, so possibly not enough to blow up. Better to create some simple tests that assert on the exact size.

jimwebber replied:
Yup. Fixed.

jimwebber committed:
Prune the Term and Index in-memory storage.

Focussing the test on actual array size (aka memory use) rather than
waiting for any potential OOM.

martinfurmanski commented:
superseded by https://github.com/neo4j/neo4j/pull/7589
